File: src/backend/decoders/bittorrent.py
def _decode_peer_id(peer_id_bytes):
    try:
        text = peer_id_bytes.decode("latin-1", errors="ignore")
    except Exception:
        return ""
    cleaned = "".join(ch if 32 <= ord(ch) <= 126 else "." for ch in text)
    return cleaned.strip(".") or ""


def decodeBitTorrent(rawPayload):
    """
    Decode common BitTorrent payload formats.

    Supports:
      - TCP handshake messages
      - TCP peer-wire length-prefixed messages
      - UDP DHT KRPC bencoded messages
    """

    PEER_WIRE_TYPES = {
        0: "choke",
        1: "unchoke",
        2: "interested",
        3: "not interested",
        4: "have",
        5: "bitfield",
        6: "request",
        7: "piece",
        8: "cancel",
        9: "port",
        20: "extended",
    }

    try:
        if not rawPayload:
            return None

        if len(rawPayload) >= 68 and rawPayload[0] == 19:
            proto = rawPayload[1:20]
            if proto == b"BitTorrent protocol":
                reserved = rawPayload[20:28]
                info_hash = rawPayload[28:48]
                peer_id = rawPayload[48:68]
                decoded_peer_id = _decode_peer_id(peer_id)

                result = {
                    "Type": "Handshake",
                    "bittorrent.type": "handshake",
                    "Protocol": "BitTorrent protocol",
                    "bittorrent.protocol": "BitTorrent protocol",
                    "Reserved": reserved.hex(),
                    "bittorrent.reserved": reserved.hex(),
                    "Info Hash": info_hash.hex(),
                    "bittorrent.info_hash": info_hash.hex(),
                    "Peer ID Hex": peer_id.hex(),
                    "bittorrent.peer_id_hex": peer_id.hex(),
                    "bittorrent.signature": "handshake",
                }
                if decoded_peer_id:
                    result["Peer ID"] = decoded_peer_id
                    result["bittorrent.peer_id"] = decoded_peer_id
                return result

        if len(rawPayload) >= 4:
            messageLength = int.from_bytes(rawPayload[0:4], "big", signed=False)
            if messageLength == 0:
                return {
                    "Type": "Peer Wire",
                    "bittorrent.type": "peer_wire",
                    "Message": "keepalive",
                    "bittorrent.message": "keepalive",
                    "Message Length": 0,
                    "bittorrent.length": 0,
                    "bittorrent.signature": "peer_wire",
                }
            if 1 <= messageLength <= len(rawPayload) - 4:
                messageId = int(rawPayload[4])
                messageName = PEER_WIRE_TYPES.get(messageId, f"id_{messageId}")
                return {
                    "Type": "Peer Wire",
                    "bittorrent.type": "peer_wire",
                    "Message": messageName,
                    "bittorrent.message": messageName,
                    "Message ID": int(messageId),
                    "bittorrent.message_id": int(messageId),
                    "Message Length": int(messageLength),
                    "bittorrent.length": int(messageLength),
                    "bittorrent.signature": "peer_wire",
                }

        if len(rawPayload) >= 10:
            text = rawPayload[:256].decode(errors="ignore")
            if text.startswith("d") and ("1:y1:" in text and ("1:q" in text or "1:r" in text)):
                y_value = ""
                y_token = "1:y1:"
                y_pos = text.find(y_token)
                if y_pos >= 0 and y_pos + len(y_token) < len(text):
                    y_value = text[y_pos + len(y_token)]
                query_name = ""
                q_token = "1:q"
                q_pos = text.find(q_token)
                if q_pos >= 0:
                    suffix = text[q_pos + len(q_token) :]
                    colon_pos = suffix.find(":")
                    if colon_pos > 0:
                        try:
                            name_len = int(suffix[:colon_pos])
                            candidate = suffix[colon_pos + 1 : colon_pos + 1 + name_len]
                            if candidate:
                                query_name = candidate
                        except Exception:
                            query_name = ""

                result = {
                    "Type": "DHT KRPC",
                    "bittorrent.type": "dht",
                    "Transaction Type": y_value or "unknown",
                    "bittorrent.transaction_type": y_value or "unknown",
                    "bittorrent.signature": "dht",
                }
                if query_name:
                    result["Query"] = query_name
                    result["bittorrent.query"] = query_name
                return result

        return None
    except Exception:
        return None

File: src/ui/decoders/conv/bittorrent.js
// BitTorrent Conv decoder: handles three BitTorrent payload shapes:
//   1. Handshake (length 68, leading 0x13, then "BitTorrent protocol")
//   2. Peer Wire (4-byte big-endian length + message ID)
//   3. DHT KRPC bencode (e.g. "d1:..." with "1:y1:q"/"1:q..." queries)

const { bytesToHexLower } = require("./smb-helpers");

const BITTORRENT_PEER_WIRE_MESSAGES = {
    0: "choke",
    1: "unchoke",
    2: "interested",
    3: "not interested",
    4: "have",
    5: "bitfield",
    6: "request",
    7: "piece",
    8: "cancel",
    9: "port",
    20: "extended",
};

function decodeBittorrentHandshake(bytes) {
    const protocol = new TextDecoder("utf-8", { fatal: false }).decode(bytes.slice(1, 20));
    if (protocol !== "BitTorrent protocol") return null;
    const infoHash = bytesToHexLower(bytes.slice(28, 48));
    const peerIdBytes = bytes.slice(48, 68);
    const peerIdHex = bytesToHexLower(peerIdBytes);
    const peerId = Array.from(peerIdBytes, (value) =>
        value >= 0x20 && value <= 0x7e ? String.fromCharCode(value) : ".",
    )
        .join("")
        .replace(/^\.+|\.+$/g, "");
    const fields = [
        { name: "Type", value: "Handshake" },
        { name: "Protocol", value: "BitTorrent protocol" },
        { name: "Info Hash", value: infoHash },
        { name: "Peer ID Hex", value: peerIdHex },
    ];
    if (peerId) fields.push({ name: "Peer ID", value: peerId });
    return { protocol: "BitTorrent", fields };
}

function decodeBittorrentPeerWire(bytes) {
    const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    const messageLength = view.getUint32(0, false);
    if (messageLength === 0) {
        return {
            protocol: "BitTorrent",
            fields: [
                { name: "Type", value: "Peer Wire" },
                { name: "Message", value: "keepalive" },
                { name: "Message Length", value: "0" },
            ],
        };
    }
    if (messageLength < 1 || messageLength > bytes.length - 4) return null;
    const messageId = bytes[4];
    const messageName = BITTORRENT_PEER_WIRE_MESSAGES[messageId] || `id_${messageId}`;
    return {
        protocol: "BitTorrent",
        fields: [
            { name: "Type", value: "Peer Wire" },
            { name: "Message", value: messageName },
            { name: "Message ID", value: String(messageId) },
            { name: "Message Length", value: String(messageLength) },
        ],
    };
}

function decodeBittorrentDhtKrpc(bytes) {
    const text = new TextDecoder("utf-8", { fatal: false }).decode(bytes.slice(0, 256));
    if (!(text.startsWith("d") && text.includes("1:y1:") && (text.includes("1:q") || text.includes("1:r")))) {
        return null;
    }
    const txMatch = text.match(/1:y1:([qre])/);
    const queryMatch = text.match(/1:q(\d+):([a-z_]+)/i);
    const fields = [
        { name: "Type", value: "DHT KRPC" },
        { name: "Transaction Type", value: txMatch?.[1] || "unknown" },
    ];
    if (queryMatch?.[2]) fields.push({ name: "Query", value: queryMatch[2] });
    return { protocol: "BitTorrent", fields };
}

function decodeBittorrentFromBytes(bytes) {
    if (!(bytes instanceof Uint8Array) || bytes.length === 0) return null;

    try {
        if (bytes.length >= 68 && bytes[0] === 19) {
            const result = decodeBittorrentHandshake(bytes);
            if (result) return result;
        }

        if (bytes.length >= 4) {
            const result = decodeBittorrentPeerWire(bytes);
            if (result) return result;
        }

        return decodeBittorrentDhtKrpc(bytes);
    } catch {
        return null;
    }
}

module.exports = { decodeBittorrentFromBytes };
